Generic ModelStateDictionary add and remove extensions

Muchiachio opened the pull request:
This adds generic add model error and remove extension methods for ModelStateDictionary, to avoid using hard coded strings then specifying model state dictionary keys. #3164

There are few concerns what comes to mind
- Initial [thought](https://github.com/aspnet/Mvc/issues/3164#issuecomment-142053478) by @rynowak was to add this to `Microsoft.AspNet.Mvc.Core`, but I couldn't because of circular references, so for now it's in `Microsoft.AspNet.Mvc.ViewFeatures`.
- Will three method be enough for covering most of the model state dictionary usage cases?
- Does `Remove<TModel>(this ModelStateDictionary, Expression)` is clear enough, maybe extension method should be named RemoveKey?
- Is adding and removing empty key on expressions without normal text representation is correct and naturally expected behavior?


<!---
@huboard:{"order":3028.794921875,"milestone_order":3182,"custom_state":""}
-->

Eilon commented:
@Muchiachio that was fast, thanks for the PR!

@dougbu can you review this?

dougbu commented:
@Eilon yes

dougbu commented on src/Microsoft.AspNet.Mvc.ViewFeatures/ModelStateDictionaryExtensions.cs:
```diff
@@ -0,0 +1,76 @@
+        /// </returns>
+        public static bool Remove<TModel>(this ModelStateDictionary modelState, Expression<Func<TModel, object>> expression)
+        {
+            return modelState.Remove(GetExpressionText(expression));
```
This choice looks incorrect. With a POCO model for example, will remove only the entry for the top-level model and not for any bound property. Should use an approach similar to `ModelBindingHelper.ClearValidationStateForModel()` or perhaps provide both `Remove()` and `RemoveAll()` methods.

Muchiachio replied:
> will remove only the entry for the top-level model and not for any bound property.

Could you elaborate a bit more on this scenario? Because I couldn't completely understand what you meant here.

dougbu replied:
Say the current model type is
``` c#
public class MyModel
{
   int A { get; set; }
   int B { get; set; }
   int C { get; set; }
   int D { get; set; }
}
```
Pass `model => model` to this method and it will remove only the `ModelStateDictionary` entry with `Key == string.Empty`. It ignores entries with names "A", "B", "C" and "D".

I'm suggesting users need an option to clean up everything for the given expression. `RemoveAll()` would provide this.

dougbu commented on src/Microsoft.AspNet.Mvc.ViewFeatures/ModelStateDictionaryExtensions.cs:
```diff
@@ -0,0 +1,76 @@
+            var unaryExpression = expression.Body as UnaryExpression;
+
+            if (IsConvertToObject(unaryExpression))
+                return ExpressionHelper.GetExpressionText(Expression.Lambda(unaryExpression.Operand, expression.Parameters[0]));
```
Braces around this expression block please.

dougbu commented on src/Microsoft.AspNet.Mvc.ViewFeatures/ModelStateDictionaryExtensions.cs:
```diff
@@ -0,0 +1,76 @@
+        /// <param name="expression">The expression of <typeparamref name="TModel"/> from
+        /// which text representation will be used as a key of the <see cref="ModelState"/> to add errors to.</param>
+        /// <param name="errorMessage"></param>
+        public static void AddModelError<TModel>(this ModelStateDictionary modelState, Expression<Func<TModel, object>> expression, string errorMessage)
```
Please use the style of most other classes in this repo: wrap parameters on their own lines when the method declaration would otherwise be significantly longer than 120 chars.

dougbu commented on src/Microsoft.AspNet.Mvc.ViewFeatures/ModelStateDictionaryExtensions.cs:
```diff
@@ -0,0 +1,76 @@
+        /// <param name="modelState">The <see cref="ModelStateDictionary"/> instance this method extends.</param>
+        /// <param name="expression">The expression of <typeparamref name="TModel"/> from
+        /// which text representation will be used as a key of the <see cref="ModelState"/> to add errors to.</param>
+        /// <param name="errorMessage"></param>
```
Don't invent new terms for this documentation, especially not "text key representation". Suggest being consistent with what we have in https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/IHtmlHelperOfT.cs That includes simply saying `specified <paramref name="expression"/>` and documenting the parameter as `An expression to be evaluated against an item in the current model.`. (Realize the expression won't be executed but it is "evaluated" in a very general sense.)

dougbu commented on test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ModelStateDictionaryExtensionsTest.cs:
```diff
@@ -0,0 +1,241 @@
+    public class ModelStateDictionaryExtensionsTest
+    {
+        [Fact]
+        public void AddsModelErrorMessage_ForSingleExpression()
```
We try to include expectations in test names. So here the name should be `AddModelError_ForSingleExpression_AddsExpectedMessage`.

dougbu commented on src/Microsoft.AspNet.Mvc.ViewFeatures/ModelStateDictionaryExtensions.cs:
```diff
@@ -0,0 +1,76 @@
+
+            return ExpressionHelper.GetExpressionText(expression);
+        }
+        private static bool IsConvertToObject(UnaryExpression expression)
```
Add a blank line above this.

dougbu replied:
Also rename to `IsConvertibleToObject`

dougbu commented:
:watch:
Note CI checks are unrelated to your changes. Should settle down once we update the MyGet feed.

Muchiachio commented:
@dougbu I updated PR with your comments.
I took similar approach to `ModelBindingHelper.ClearValidationStateForModel`, but empty key would not delete all keys, just the model property ones.

dougbu commented on src/Microsoft.AspNet.Mvc.ViewFeatures/ModelStateDictionaryExtensions.cs:
```diff
@@ -0,0 +1,115 @@
+                if (elementMetadata != null)
+                {
+                    modelMetadata = elementMetadata;
+                }
```
I can't think of a case where property names of an element will appear in the `ModelStateDictionary` at the top level. A top-level collection of simple elements e.g. `string[]` might be bound as `string.Empty` (though HTML generation disallows the empty key) but will normally appear with prefixes such as `"[0]"`.

Suggest removing this block for now. Worst case is the block below will attempt to clean out entries with the `"Count"`, `"IsReadOnly"` or similar prefix. File a follow-up issue to find all collection prefixes and do the right thing. For example, we could eventually find the collection prefixes borrowing code from `JQueryFormValueProviderFactory.NormalizeJQueryToMvc()`. No incredible rush -- top-level collections are unusual.

dougbu commented on src/Microsoft.AspNet.Mvc.ViewFeatures/ModelStateDictionaryExtensions.cs:
```diff
@@ -0,0 +1,115 @@
+        private static bool IsConvertibleToObject(UnaryExpression expression)
+        {
+            return expression?.NodeType == ExpressionType.Convert &&
+                expression.Operand is MemberExpression &&
```
`expression.Operand?.NodeType == ExpressionType.MemberAccess`

dougbu commented on src/Microsoft.AspNet.Mvc.ViewFeatures/ModelStateDictionaryExtensions.cs:
```diff
@@ -0,0 +1,115 @@
+            return ExpressionHelper.GetExpressionText(expression);
+        }
+
+        private static bool IsConvertibleToObject(UnaryExpression expression)
```
Sorry, my name suggestion didn't make sense. Prefer `IsConversionToObject()`.

dougbu commented on src/Microsoft.AspNet.Mvc.ViewFeatures/ModelStateDictionaryExtensions.cs:
```diff
@@ -0,0 +1,115 @@
+
+        private static string GetExpressionText(LambdaExpression expression)
+        {
+            var unaryExpression = expression.Body as UnaryExpression;
```
Add code comments informing reader that this is about ignoring an outermost conversion to `object`.

Forgot to ask last time: What scenarios hit this case? We don't have similar handling elsewhere. If it's an important scenario, probably need to handle it at a lower level instead.

Muchiachio replied:
It's all about not having to specify to two generic types then calling an extension method, so we don't have to have boilerplate code like this:
`ModelState.AddModelError<Account, *int*>(model => model.Age, "You are too young.")`
And instead can write
`ModelState.AddModelError<Account>(model => model.Age, "You are too young.")`
which on structs result in automatic cast to object, because
`Expression<Func<TModel, object>> expression` converts stuff like `model => model.Age` to `model => Convert(model.Age)` which can not be converted to proper text by `ExpressionHelper.GetExpressionText` method.

dougbu replied:
If this is needed only when the expression starts with `(object)` explicitly, please remove. If it's required because Linq generates a `Convert` node for a boxing operation or other implicit conversion, :ok:

Muchiachio replied:
Yes, Linq generates a `Convert` node. I will add a comment describing an issue this code block is solving.

dougbu commented on test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ModelStateDictionaryExtensionsTest.cs:
```diff
@@ -0,0 +1,336 @@
+            modelState.AddModelError<TestModel>(model => model.Text, "Message");
+
+            // Assert
+            Assert.Equal("Text", modelState.Single().Key);
```
FYI only: `Assert.Single()` results in significantly better test failures than the `Exception` Linq's `Single()` throws. If you're kind enough to tackle other issues in the future, do this as `Assert.Equal("Text", Assert.Single(modelState).Key);`. Or use a temporary variable for the `Assert.Single()` result so that each assertion is on a single line and the exact failure is easier to find.

Muchiachio replied:
Definitely will change that to `Assert.Single` :+1:

dougbu commented on test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ModelStateDictionaryExtensionsTest.cs:
```diff
@@ -0,0 +1,336 @@
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            modelState.AddModelError<TestModel>(model => (object)model.Child.Value, exception);
```
Note this expression is not supported elsewhere in MVC. The leftmost expression can't be a cast.

Muchiachio replied:
Check my comment above, about automatic struct casting to object. I just wanted to explicitly add this, so that it would be visible. I can remove the cast if you think it's okay.

dougbu replied:
Suggest removing the tests involving explicit casts and `ToString()`. The implicit cast test seems sufficient.

dougbu commented on test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ModelStateDictionaryExtensionsTest.cs:
```diff
@@ -0,0 +1,336 @@
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            modelState.AddModelError<TestModel>(model => model.ToString(), exception);
```
This expression is also not supported in MVC.

dougbu commented on test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ModelStateDictionaryExtensionsTest.cs:
```diff
@@ -0,0 +1,336 @@
+    public class ModelStateDictionaryExtensionsTest
+    {
+        [Fact]
+        public void AddModelError_ForSingleExpression_AddsExpectedMessage()
```
Optional but may be worth testing field or variable references in the expression e.g.
``` c#
  var something = "whatever value";
  modelState.AddModelError<TestModel>(model => something, ...);
```
The above would add an error with key "something" and it's a supported case.

dougbu commented:
@Muchiachio looking quite good. The current checks appear to be failing for reasons unrelated to your changes -- should clear up soon.

dougbu commented:
:watch:

Muchiachio commented:
I updated the code based on our conversation, and replaced explicit and `.ToString()` tests to variable reference ones. Seems like I didn't get a CLA check.

dougbu commented:
Cycling closed / open to see if the CLA bot wakes up...

dnfclas commented:
Hi __@Muchiachio__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

dougbu commented:
:shipit: I'll get this in soon.

dougbu commented:
squashed to b28debf with slight touch-up in eff10cd (move to same namespace as `ModelStateDictionary`)
